tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,301 @@
+                    be set to a value not equal to 1. Otherwise, you will have the exact same
+                    result as not using discriminative learning.
+
+                learning_rate: float, the learning rate for the model
+
+                verbose: Bool, to generate a report on how many parameters are affected
+
+                *args: Args to pass to the base optimizer
+
+                **kwargs: Kwargs to pass to the base optimizer
+
+            Returns
+                Optimizer - A keras optimizer to use with model.compile
+
+            References
+                - [Universal Language Model Fine-tuning for Text Classification](https://arxiv.org/pdf/1801.06146.pdf)
+        """
+        assert issubclass(
+            base_optimizer, tf.keras.optimizers.Optimizer
+        ), "Base optimizer must be a class that inherits from tf.keras.optimizers.Optimizer"
+
+        super().__init__(lr=learning_rate, name=name, *args, **kwargs)
+
+        DiscriminativeModelManager._prepare_model(model, verbose=verbose)
+
+        self.opt_class = base_optimizer
+        self.learning_rate = learning_rate
+        self.kwargs = kwargs
+
+        # find unique lr_mult
+        variable_groups = {var.lr_mult: None for var in model.trainable_variables}
+
+        self.optimizer_group = []
+
+        for lr_mult in variable_groups.keys():
+            opt = self.opt_class(learning_rate=learning_rate * lr_mult, **kwargs)
+            opt.lr_mult = lr_mult
+            self.optimizer_group.append(opt)
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        """allocates gradients to each optimizer based on the variable's learning rate multiplier
+        then applies the gradients. In graph mode, it returns 1 operation per optimizer
+        Please use the model.fit method instead of accessing this directly
+        """
+
+        # create gradvar buckets for each opt
+        gvdict = {}
+        for opt in self.optimizer_group:
+            gvdict[opt.lr_mult] = []
+
+        # load the gradvars into the appropriate bucket
+        for grad, var in tuple(grads_and_vars):
+            gvdict[var.lr_mult].append((grad, var))
+
+        # return results from each opt
+        # in eager mode, this will return a list of irrelevant results for each optimizer
+        # in eager mode, the function apply_gradients actually applies gradients to the model
+        # in graph mode, this will return a list of tensor ops for each opt
+        # in graph mode, apply_gradients creates the tensor ops for applying gradients on the graph
+        return [
+            opt.apply_gradients(tuple(gvdict[opt.lr_mult]))
+            for opt in self.optimizer_group
+        ]
+
+    def get_config(self):
+        """This method cannot effectively return the optimizer configuration because
+        that configuration depends on the model and base optimizer
+        for now, it returns the config values of itself and base optimizers
+        """
```

Comment:
This might be an issue when trying to save/load a model.

Reply:
I can investigate this further, but this type of optimizer is directly linked to the model. I'm not sure if it would be appropriate to return the model when getting the config of the optimizer. 

The alternative is to store the model parameters in the optimizer or at least the model parameter names and their assigned lr_mult. This could make the implementation more complex.

Reply:
I believe I misread the get/from_config class methods of the optimizerv2. It seems that it does not save the optimizer state. See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L751

Reply:
I misunderstood the code logic. Initially, I thought you were passing an object of the base optimizer, not the class. You're `get_config`, `from_config` seem fine.